src/Preact10ShallowDiff.ts

```diff
@@ -0,0 +1,759 @@
+
+      if (renderResult == skipUpdateSymbol) {
+        // The memo component told us not to update, so return the result of the
+        // previous render
+        return this._rendered;
+      }
+
+      if (isComponentVNode(renderResult)) {
+        // The memo component updated, so let's setup to update and diff the
+        // component it returned
+        newVNode = renderResult;
+        oldVNode = this._prevElement ?? ({} as any);
+      } else {
+        throw new Error(
+          `Memo rendered a non-component element. Only memo'ed components is currently supported for shallow rendering`
+        );
+      }
+    }
+
+    this._prevElement = newVNode;
+    renderResult = diffComponent(
+      newVNode,
+      oldVNode,
+      globalContext,
+      commitQueue
+    );
+
+    // If diffing this component told us not to update, then return the result
+    // of the previous render. Else, return the result of the diff
+    return renderResult === skipUpdateSymbol ? this._rendered : renderResult;
+  }
+}
+
+function isComponentVNode<P = any>(element: any): element is ComponentVNode<P> {
+  return isValidElement(element) && typeof element.type === 'function';
+}
+
+function assertIsComponentVNode(
+  element: JSX.Element
+): asserts element is ComponentVNode {
+  if (!isValidElement(element)) {
+    throw new Error(
+      `Preact10ShallowDiff render(): Invalid component element. ${
+        typeof element === 'function'
+          ? 'Instead of passing a component class, make sure to instantiate it by passing it to Preact.createElement.'
+          : ''
+      }`
+    );
+  }
+
+  // Show a special message for host elements since it's a common case.
+  if (typeof element.type === 'string') {
+    throw new Error(
+      `Preact10ShallowDiff render(): Shallow rendering works only with custom components, not primitives (${element.type}). Instead of calling \`.render(el)\` and inspecting the rendered output, look at \`el.props\` directly instead.`
+    );
+  }
+
+  if (typeof element.type !== 'function') {
+    throw new Error(
+      `Preact10ShallowDiff render(): Shallow rendering works only with custom components, but the provided element type was \`${
+        Array.isArray(element.type)
+          ? 'array'
+          : element.type === null
+          ? 'null'
+          : typeof element.type
+      }\`.`
+    );
+  }
+}
+
+/** Determine if a VNode is a Memo wrapper */
+function isMemo(node: VNode) {
+  if (node?.type && typeof node.type === 'function') {
+    const nodeType = node.type;
+    return (
+      (nodeType as any)._forwarded === true &&
+      nodeType.displayName?.startsWith('Memo(')
+    );
+  }
+
+  return false;
+}
+
+/** Symbol to return if a component indicates it should not update */
+const skipUpdateSymbol = Symbol('Preact10ShallowDiff skip update');
+
+/**
+ * Shallowly diff a component. Much of this function is copied directly from the
+ * Preact 10 source. Differences to the original source are commented.
```

Comment:
If we put the shallow renderer in a separate directory, then I think it would also make sense to split `diffComponent` out into a separate file, along with helpers that were also copied from the Preact source code. We might need to include or at least indicate the license at the top, if different from the rest of the code.